File: latest/week3/workshop7_m4/W07_M04_demo.R
# ============================================================
# Workshop 7 — Module 4: The Seven Basic Data Stories
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ── LOAD BOTH DATASETS ──────────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")))
nf_genres <- nf |> separate_rows(listed_in, sep = ", ") |> mutate(genre = str_trim(listed_in))

ec <- read_csv("ecar.csv") |>
  mutate(Year = year(mdy(`Approve Date`)), Spread = Rate - `Cost of Funds`) |>
  filter(Year >= 2002, Year <= 2012)

# ═══════════════════════════════════════════════════════
# STORY 1: CHANGE OVER TIME
# Finding: "Movie additions peaked in 2019 and declined 42%"
# ═══════════════════════════════════════════════════════
yearly <- nf |> filter(!is.na(year_added), year_added >= 2015, year_added <= 2021) |>
  count(year_added, type)
movies <- yearly |> filter(type == "Movie")
peak <- movies |> slice_max(n, n = 1)
decline <- round((1 - movies$n[movies$year_added == 2021] / peak$n) * 100)

p1 <- ggplot(yearly, aes(x = year_added, y = n, group = type)) +
  geom_line(color = "#DDDDDD", linewidth = 0.8) +
  geom_line(data = movies, color = "#1565C0", linewidth = 2.5) +
  annotate("label", x = 2018, y = movies$n[movies$year_added == 2021] * 0.85,
    label = paste0("–", decline, "% from peak"),
    size = 3, fontface = "bold", color = "#E53935", fill = "white", label.size = 0.4) +
  geom_text(data = yearly |> filter(year_added == 2021),
    aes(label = type, color = type == "Movie"),
    hjust = -0.1, size = 2.5, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#BBB")) +
  coord_cartesian(xlim = c(2015, 2022.5)) +
  theme_minimal(base_size = 8) + theme(legend.position = "none") +
  labs(title = "STORY 1 — Change Over Time:\n'Movie additions peaked in 2019 and declined since'",
       y = "Titles Added")

# ═══════════════════════════════════════════════════════
# STORY 2: DRILL DOWN
# Finding: "69% movies, 31% TV shows"
# ═══════════════════════════════════════════════════════
type_split <- nf |> count(type) |> mutate(pct = n / sum(n) * 100)
p2 <- ggplot(type_split, aes(x = fct_reorder(type, -pct), y = pct, fill = type)) +
  geom_col(width = 0.5, show.legend = FALSE) +
  geom_text(aes(label = paste0(round(pct), "%")), vjust = -0.5, size = 3, fontface = "bold") +
  scale_fill_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal(base_size = 8) +
  labs(title = "STORY 2 — Drill Down:\n'69% movies, 31% TV shows'", y = "%", x = NULL)

# ═══════════════════════════════════════════════════════
# STORY 3: ZOOM OUT
# Finding: "India rose from #5 to #2 globally"
# ═══════════════════════════════════════════════════════
country_yr <- nf |> filter(!is.na(primary_country), !is.na(year_added),
  year_added >= 2016, year_added <= 2021) |>
  count(year_added, primary_country) |>
  group_by(primary_country) |> arrange(year_added) |> mutate(cum = cumsum(n)) |> ungroup()

top5c <- country_yr |> filter(year_added == 2021) |> slice_max(cum, n = 5) |> pull(primary_country)
country_top5 <- country_yr |> filter(primary_country %in% top5c) |>
  group_by(year_added) |> mutate(rank = rank(-cum, ties.method = "first")) |> ungroup()

p3 <- ggplot(country_top5, aes(x = year_added, y = rank, color = primary_country)) +
  geom_line(linewidth = 1.2) + geom_point(size = 2) +
  scale_y_reverse(breaks = 1:5) +
  geom_text(data = country_top5 |> filter(year_added == 2021),
    aes(label = primary_country), hjust = -0.15, size = 2.5, fontface = "bold",
    show.legend = FALSE) +
  coord_cartesian(xlim = c(2016, 2022.5)) +
  theme_minimal(base_size = 8) + theme(legend.position = "none") +
  labs(title = "STORY 3 — Zoom Out:\n'India rose from #5 to #2 in global content'",
       y = "Rank (1=most)")

# ═══════════════════════════════════════════════════════
# STORY 4: CONTRAST
# Finding: "Pre-2008 spread: 3.8pp vs Post-2008: 2.9pp"
# ═══════════════════════════════════════════════════════
ec_period <- ec |> mutate(period = ifelse(Year < 2008, "Pre-2008", "Post-2008"))
spread_compare <- ec_period |> group_by(period) |>
  summarise(med_spread = median(Spread), .groups = "drop")

p4 <- ggplot(spread_compare, aes(x = period, y = med_spread, fill = period)) +
  geom_col(width = 0.4, show.legend = FALSE) +
  geom_text(aes(label = paste0(round(med_spread, 1), "pp")),
    vjust = -0.5, size = 3.5, fontface = "bold") +
  scale_fill_manual(values = c("Pre-2008" = "#1565C0", "Post-2008" = "#E53935")) +
  theme_minimal(base_size = 8) +
  labs(title = "STORY 4 — Contrast:\n'Spread compressed from 3.8pp to 2.9pp after crisis'",
       y = "Median Spread (pp)", x = NULL)

# ═══════════════════════════════════════════════════════
# STORY 5: INTERSECTIONS
# Finding: "Higher tier → lower rate"
# ═══════════════════════════════════════════════════════
p5 <- ggplot(ec |> sample_n(2000), aes(x = factor(Tier), y = Rate)) +
  geom_boxplot(fill = "#BBDEFB", color = "#1565C0", width = 0.5, outlier.size = 0.3) +
  theme_minimal(base_size = 8) +
  labs(title = "STORY 5 — Intersections:\n'Higher tier → lower rate (strong relationship)'",
       x = "Credit Tier", y = "Rate (%)")

# ═══════════════════════════════════════════════════════
# STORY 6: FACTORS
# Finding: "Three factors drive genre growth"
# ═══════════════════════════════════════════════════════
factors_df <- tibble(
  factor = c("Global expansion\n(2016, 130 countries)", "Original production\ninvestment",
    "Local-language\ncontent", "Licensing deals", "COVID demand surge"),
  impact = c(35, 28, 22, 10, 5)) |>
  mutate(factor = fct_reorder(factor, impact))

p6 <- ggplot(factors_df, aes(x = factor, y = impact)) +
  geom_col(fill = "#00695C", width = 0.5) +
  geom_text(aes(label = paste0(impact, "%")), hjust = -0.2, size = 3, fontface = "bold",
    color = "#00695C") +
  coord_flip(clip = "off") +
  theme_minimal(base_size = 8) +
  labs(title = "STORY 6 — Factors:\n'Three drivers explain Netflix genre growth'",
       x = NULL, y = "Estimated Impact (%)")

# ═══════════════════════════════════════════════════════
# STORY 7: OUTLIERS
# Finding: "Dec 2020: 160 titles — 3× average"
# ═══════════════════════════════════════════════════════
monthly <- nf |> filter(!is.na(date_added), year(date_added) >= 2016) |>
  mutate(month = floor_date(date_added, "month")) |> count(month)
avg <- mean(monthly$n)
monthly <- monthly |> mutate(is_outlier = n > avg * 2)

p7 <- ggplot(monthly, aes(x = month, y = n, color = is_outlier)) +
  geom_point(size = 1.5) + geom_line(color = "#DDDDDD", linewidth = 0.3) +
  scale_color_manual(values = c("TRUE" = "#E53935", "FALSE" = "#BBBBBB"), guide = "none") +
  geom_hline(yintercept = avg, linetype = "dashed", color = "#888") +
  geom_hline(yintercept = avg * 2, linetype = "dotted", color = "#E53935") +
  annotate("text", x = as.Date("2017-01-01"), y = avg * 2 + 5,
    label = "2× average threshold", size = 2, color = "#E53935") +
  theme_minimal(base_size = 8) +
  labs(title = "STORY 7 — Outliers:\n'Several months exceeded 2× the average additions'",
       x = NULL, y = "Titles Added")

# ═══════════════════════════════════════════════════════
# COMPOSE 7-STORY DASHBOARD
# ═══════════════════════════════════════════════════════
dashboard <- (p1 | p2 | p3) / (p4 | p5) / (p6 | p7) +
  plot_annotation(
    title = "Seven Data Stories Applied to Netflix + e-Car",
    subtitle = "Each finding maps to a story type → story type maps to a chart",
    tag_levels = "1")
ggsave("output/seven_stories_applied.png", dashboard, width = 16, height = 14, dpi = 300)

# Individual saves
ggsave("output/story1_change.png", p1, width = 8, height = 4, dpi = 300)
ggsave("output/story4_contrast.png", p4, width = 6, height = 4, dpi = 300)
ggsave("output/story7_outliers.png", p7, width = 8, height = 4, dpi = 300)

cat("\n── All W07-M04 R outputs saved ──\n")

File: latest/week3/workshop7_m4/W07_M04_demo.py
"""W07-M04: The Seven Basic Data Stories — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.gridspec import GridSpec
os.makedirs("output", exist_ok=True)

nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
nf["primary_country"] = nf["country"].str.split(",").str[0].str.strip()

ec = pd.read_csv("ecar.csv"); ec.columns=[c.strip().replace("  "," ") for c in ec.columns]
ec["Year"] = pd.to_datetime(ec["Approve Date"], format="%m/%d/%Y", errors="coerce").dt.year
ec["Spread"] = ec["Rate"] - ec["Cost of Funds"]
ec = ec.query("2002<=Year<=2012")

yearly = nf.query("2015<=year_added<=2021").groupby(["year_added","type"]).size().reset_index(name="n")
movies = yearly.query("type=='Movie'"); tvshows = yearly.query("type=='TV Show'")

# 7-PANEL DASHBOARD
fig = plt.figure(figsize=(16, 14)); gs = GridSpec(3, 3, figure=fig, hspace=0.45, wspace=0.35)

# S1: CHANGE OVER TIME
ax = fig.add_subplot(gs[0,0])
ax.plot(tvshows.year_added, tvshows.n, "-o", color="#DDD", lw=0.8, markersize=3)
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2, markersize=5)
peak_n = movies.n.max(); latest_n = movies.iloc[-1]["n"]
ax.annotate(f"–{round((1-latest_n/peak_n)*100)}%", xy=(2020, (peak_n+latest_n)/2),
    fontsize=9, fontweight="bold", color="#E53935", ha="center",
    bbox=dict(boxstyle="round,pad=0.2", facecolor="white", edgecolor="#E53935"))
ax.set_title("S1: Change Over Time\n'Movies peaked 2019, declined since'", fontsize=8, fontweight="bold", color="#1565C0")

# S2: DRILL DOWN
ax = fig.add_subplot(gs[0,1])
type_split = nf["type"].value_counts()
ax.barh(type_split.index, type_split.values/type_split.sum()*100, color=["#1565C0","#E53935"], height=0.4)
for i,(t,v) in enumerate(zip(type_split.index, type_split.values/type_split.sum()*100)):
    ax.text(v+1, i, f"{v:.0f}%", va="center", fontsize=8, fontweight="bold")
ax.set_title("S2: Drill Down\n'69% movies, 31% TV'", fontsize=8, fontweight="bold", color="#2E7D32")

# S3: ZOOM OUT (bump chart)
ax = fig.add_subplot(gs[0,2])
cyr = (nf.dropna(subset=["primary_country"]).query("2016<=year_added<=2021")
    .groupby(["year_added","primary_country"]).size().reset_index(name="n"))
cyr["cum"] = cyr.sort_values("year_added").groupby("primary_country")["n"].cumsum()
top5c = cyr.query("year_added==2021").nlargest(5,"cum")["primary_country"].tolist()
for c in top5c:
    sub = cyr.query("primary_country==@c").sort_values("year_added")
    ranks = sub.groupby("year_added")["cum"].rank(ascending=False, method="first")
    ax.plot(sub.year_added, ranks, "-o", lw=1, markersize=3)
ax.set_ylim(5.5, 0.5); ax.set_title("S3: Zoom Out\n'India rose to #2 globally'", fontsize=8, fontweight="bold", color="#E65100")

# S4: CONTRAST
ax = fig.add_subplot(gs[1,0])
pre = ec.query("Year<2008")["Spread"].median(); post = ec.query("Year>=2008")["Spread"].median()
ax.bar(["Pre-2008","Post-2008"], [pre, post], color=["#1565C0","#E53935"], width=0.4)
ax.set_title(f"S4: Contrast\n'Spread: {pre:.1f}pp → {post:.1f}pp'", fontsize=8, fontweight="bold", color="#C62828")
ax.set_ylabel("Median Spread (pp)")

# S5: INTERSECTIONS
ax = fig.add_subplot(gs[1,1])
sample = ec.sample(2000, random_state=42)
bp = ax.boxplot([sample.query("Tier==@t")["Rate"].values for t in sorted(ec.Tier.unique())],
    labels=[f"T{t}" for t in sorted(ec.Tier.unique())], patch_artist=True, widths=0.5)
for patch in bp["boxes"]: patch.set_facecolor("#BBDEFB")
ax.set_title("S5: Intersections\n'Higher tier → lower rate'", fontsize=8, fontweight="bold", color="#7B1FA2")
ax.set_ylabel("Rate (%)")

# S6: FACTORS
ax = fig.add_subplot(gs[1,2])
factors = ["Global expansion","Original production","Local-language","Licensing","COVID demand"]
impacts = [35, 28, 22, 10, 5]
ax.barh(factors, impacts, color="#00695C", height=0.5)
for i, v in enumerate(impacts): ax.text(v+0.5, i, f"{v}%", va="center", fontsize=7, fontweight="bold", color="#00695C")
ax.set_title("S6: Factors\n'Three drivers of genre growth'", fontsize=8, fontweight="bold", color="#00695C")

# S7: OUTLIERS
ax = fig.add_subplot(gs[2,0])
monthly = nf.dropna(subset=["date_added"]).query("date_added.dt.year>=2016").set_index("date_added").resample("ME").size()
avg = monthly.mean()
colors = ["#E53935" if v > avg*2 else "#BBBBBB" for v in monthly.values]
ax.scatter(monthly.index, monthly.values, s=10, c=colors, zorder=5)
ax.plot(monthly.index, monthly.values, color="#EEEEEE", lw=0.3)
ax.axhline(avg, ls="--", color="#888", lw=0.5); ax.axhline(avg*2, ls=":", color="#E53935", lw=0.5)
ax.set_title("S7: Outliers\n'Several months > 2× average'", fontsize=8, fontweight="bold", color="#E53935")

# Decision table panel
ax = fig.add_subplot(gs[2,1:]); ax.axis("off")
rows = [
    ("'grew / declined / peaked'", "Change Over Time", "Line chart"),
    ("'composed of / breakdown'", "Drill Down", "Stacked bar"),
    ("'compared to / in context'", "Zoom Out", "Small multiples"),
    ("'vs / gap between'", "Contrast", "Grouped bar, slope"),
    ("'correlated / related'", "Intersection", "Scatter"),
    ("'driven by / because'", "Factors", "Waterfall, tornado"),
    ("'unusual / spike / anomaly'", "Outlier", "Highlight scatter"),
]
ax.text(0.5, 0.95, "Finding → Story → Chart Decision Table", ha="center",
    fontsize=10, fontweight="bold", transform=ax.transAxes)
for i, (keyword, story, chart) in enumerate(rows):
    y = 0.82 - i * 0.11
    ax.text(0.05, y, keyword, fontsize=7, color="#555", transform=ax.transAxes)
    ax.text(0.42, y, f"→ {story}", fontsize=7, fontweight="bold", color="#333", transform=ax.transAxes)
    ax.text(0.72, y, f"→ {chart}", fontsize=7, color="#888", fontstyle="italic", transform=ax.transAxes)

for a in fig.get_axes():
    if hasattr(a, 'spines'):
        a.spines["top"].set_visible(False); a.spines["right"].set_visible(False)
    a.tick_params(labelsize=5)

fig.suptitle("Seven Data Stories Applied to Netflix + e-Car", fontsize=14, fontweight="bold")
plt.savefig("output/seven_stories_applied.png", dpi=300, bbox_inches="tight"); plt.close()

print("\nAll W07-M04 Python outputs saved")
